core/src/vote_stake_tracker.rs

```diff
@@ -1,39 +1,46 @@
 use {solana_sdk::pubkey::Pubkey, std::collections::HashSet};
 
+#[derive(Default)]
+pub struct VoteThresholdCheckResult {
+    pub check_duplicate: bool,
+    pub check_vote: bool,
+    pub is_new: bool,
+}
 #[derive(Default)]
 pub struct VoteStakeTracker {
     voted: HashSet<Pubkey>,
     stake: u64,
 }
 
 impl VoteStakeTracker {
-    // Returns tuple (reached_threshold_results, is_new) where
-    // Each index in `reached_threshold_results` is true if the corresponding threshold in the input
-    // `thresholds_to_check` was newly reached by adding the stake of the input `vote_pubkey`
-    // `is_new` is true if the vote has not been seen before
+    // Returns VoteThresholdCheckResult.
+    // It checks both DUPLICATE_THRESHOLD and VOTE_THRESHOLD_SIZE by adding the
+    // stake of the vote from the new `vote_pubkey` and comparing against the target
+    // threshold.
+    // `is_new` is true if the vote_pubkey has not been seen before.
     pub fn add_vote_pubkey(
         &mut self,
         vote_pubkey: Pubkey,
         stake: u64,
         total_stake: u64,
         thresholds_to_check: &[f64],
```

Comment:
Well this is getting a slice of threshold to check. So this can be of any length, not just 2.

If you want to hard-code this to only 2 thresholds, then we need this slice to be also updates to a struct/tuple or something with only 2 thresholds.

But maybe first confirm with @carllin that hard-coding this thresholds slice to only 2 values would be ok; or if there was any reason to define the thresholds as a slice.

Reply:
yeah. @carllin what do you think? Are we going to need more than 2 threshods?

Reply:
Yes, I believe Anatoly eventually wanted us to be able to support user-defined thresholds

Reply:
Ah. ok. I will shelve this change and keep the code as it is for the flexibility.

Reply:
To bring up a few points:

1. `add_vote_key()` is called for each vote transaction. So it is very heavily called.
2. `VoteStakeTracker` is behind a write lock. So speeding it up will 
(a) shorten the lock time that helps the reader of the stake() to process the vote. 
(b) speed up the sending of the vote. 
https://github.com/solana-labs/solana/blob/c785f1ffc5c0d85e7aa42a01011c820c4d24bbea/core/src/cluster_info_vote_listener.rs#L643

As a result, the speed up of `VoteStakeTracker` will make the vote faster and help the who cluster to reach consensus quickcer.